service/csv/csv_worker.go

```diff
@@ -0,0 +1,125 @@
+package csv
+
+import (
+	"encoding/csv"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/rodrigomkd/go-rest-api/model"
+)
+
+type ICSVWService interface {
+	ReadWorkers() []model.Worker
+}
+
+type CSVWorkerService struct {
+	csv            ICSVWService
+	dataSourcePath string
+}
+
+func NewWorker(dataSourcePath string) *CSVWorkerService {
+	return &CSVWorkerService{
+		dataSourcePath: dataSourcePath,
+	}
+}
+
+func (cs CSVWorkerService) ReadWorkers(typ string, items int, itemsPerWork int) []model.Worker {
+	f, _ := os.Open(cs.dataSourcePath)
+	defer f.Close()
+
+	ts := time.Now()
+	activities := cs.concuRSwWP(f, typ, items, itemsPerWork)
+	te := time.Now().Sub(ts)
+
+	fmt.Println("END Concurrency: ", te)
+
+	return activities
+}
+
+// with Worker pools
+func (cs CSVWorkerService) concuRSwWP(f *os.File, typ string, items int, itemsPerWork int) []model.Worker {
+	fcsv := csv.NewReader(f)
+	rs := make([]model.Worker, 0)
+	numWps := items / itemsPerWork
+	jobs := make(chan []string, numWps)
+	res := make(chan model.Worker)
+
+	var wg sync.WaitGroup
+	worker := func(jobs <-chan []string, results chan<- model.Worker) {
+		for {
+			select {
+			case job, ok := <-jobs: // must check for readable state of the channel.
```

Comment:
This doesn't really guarantee that each worker is going to be taking exactly the number of jobs they should, each one could have a counter of how many jobs have processed and exit when they have reached that limit